rsub.1
.Dd October 8, 2026
.Dt RSUB 1
.Os
.Sh NAME
.Nm rsub
.Nd substitute lines or block of text
.Sh SYNOPSIS
.Nm rsub
.Op Fl A
.Fl l Ar line_text
.Fl r Ar line_regex
.Ar target
.Nm rsub
.Ar target
.Ar < block_content
.Sh DESCRIPTION
.Nm
is shipped to remote machines by
.Xr rset 1
to provide a standard method of modifying configuration parameters within a
file.
Two modes of operation are supported: line substitution and block replacement.
.Pp
Single line replacements are defined by the flags
.Fl r
and
.Fl l ,
while the content for a text block is provided on STDIN.
.Pp
The arguments for line substitution are as follows:
.Bl -tag -width Ds
.It Fl A
Append line if the pattern is not found.
.It Fl l Ar line_text
The replacement text for a matching line.
Newlines are not permitted as part of the replacement string.
.It Fl r Ar line_regex
A regular expression matching the line to replace.
.El
.Sh ENVIRONMENT
.Bl -tag -width Ds
.It Ev RSUB_DIFF_ARGS
Arguments for
.Xr diff 1
tool.
The default is
.Qq -U 2 .
.It Ev RSUB_START RSUB_END
Define the beginning and ending markers when updating a block of text.
.El
.Sh EXIT STATUS
The
.Nm
utility exits with one of the following values:
.Pp
.Bl -tag -width Ds -offset indent -compact
.It Li 0
Target file was updated
.It Li 1
No changes were made, or an error occurred
.It Li 2
Incorrect parameters
.It Li 3
File not found
.El
.Sh EXAMPLES
Ensure a single configuration parameter is set
.Bd -literal -offset indent
$SD/rsub -r '^AllowTcpForwarding .+' -l 'AllowTcpForwarding yes' \e
	/etc/ssh/sshd_config
.Ed
.Pp
Update a managed block of text
.Bd -literal -offset indent
$SD/rsub /etc/fstab <<-CONF
/dev/ada0p1  /vm  ufs  rw  0  0
CONF
.Ed
.Sh SEE ALSO
.Xr awk 1 ,
.Xr rinstall 1
.Sh HISTORY
The
.Nm
utility was introduced in rset 0.7


NEWS
= Release History

== 3.5: October 8, 2026

 - Dynamically allocate route labels; use arbitrary limit of 10000 routes
 - Set a limit of 1000 for range expansion
 - New 'init' command with example configurations for 'alpine' and 'openbsd'
 - New 'wenv' command, use custom log format to supersede '-e' flag
 - Display multibyte characters in error messages
 - Enable environment_file= to accept multiple pathnames
 - pln(5): no longer document single-use options
 - pln(5): rewrite man page and organize in three major sections
 - renv(1): require closing quote, reject trailing whitespace

== 3.4: August 3, 2026

 - Match route labels using regular expressions, or any alias using exact hostnames
 - Do not count zero-width regex as a match
 - All utilities: add usage hint and '-h' flag to display option summary
 - Detect string copy overflow and abort with an error
 - rinstall(1) create directories using the same owner as ${SD}
 - miniquark(1) compare HTTP If-Modified-Since times using GMT
 - miniquark(1) adjust options to resemble nc(1)
 - Use 'ustar' format for transfering files on OpenBSD
 - Replace 'tmppath' pledge with 'wpath cpath' on OpenBSD
 - Exit without cleanup if LOCAL_CONTROL_SOCKET already exists
 - Wait 0.5 seconds for SSH control socket to clear

== 3.3: January 17, 2026

 - Use session_id for remote stage dir name
 - Access miniquark logs using HTTP_TRACE
 - Access ssh invocation using SSH_TRACE
 - Do not set INSTALL_URL, rely on default in rinstall(1)
 - Exit on failure for malloc(2), pipe(2), socket(2), bind(2), strdup(2)
 - Allow export path on host labels used for archive/restore
 - pln(5) Ensure export paths are only specified on route labels
 - pln(5) Raise error for indented text in unexpected context
 - miniquark(1) listen on multiple addresses, log each bind address
 - miniquark(1) remove chroot feature
 - rinstall(1) default to curl

== 3.2: February 6, 2025

 - Update MANPREFIX for FreeBSD
 - Format source using clang-format and rubocop
 - rsub(1) allow diff(1) behavior to be modified using RSUB_DIFF_ARGS
 - rsub(1) refactor based on patterns used in rinstall(1)
 - rsub(1) remove temporary source file
 - Ensure top-level directory permissions are not world-readable
 - Avoid leaving a NUL character at the end of local execution script
 - Use native strtonum(3) on MacOS

== 3.1: September 24, 2024

 - Add begin= and end= options
 - Add archive and restore options
 - Allow IPv6 addresses to be used in routes
 - Add out-of-bounds check to options
 - Use 'set -a' instead of 'sh -a' to avoid environment contamination
 - Enable parallel execution using background workers
 - rinstall(1) print name of missing source with absolute path
 - rinstall(1) adjust comments and error messages

== 3.0: February 26, 2024

 - Check for ssh-agent before updating _rutils
 - Correct use of getopt(3) to allow compact flags
 - Use RSET_HOST_CONNECT_ERROR to format failed connection or stagedir initialization
 - Use RSET_LABEL_EXEC_ERROR if ssh(1) or sh(1) terminates during script execution
 - Add log format variable '%s' for generated session ID
 - Remove '-l' option
 - Tests: use curl(1) on Linux
 - Detect and use native strlcpy(3), arc4random(3) on Linux
 - Upload _rutils and exported paths in two steps for compatibility with busybox tar(1)
 - rsub(1): escape ampersands found in '-l' argument
 - rinstall(1) allow diff(1) behavior to be modified using RINSTALL_DIFF_ARGS
 - rinstall(1) add '-a' option for defining an alternate source location
 - rinstall(1) saving a file in $SD no longer depends the current directory
 - rinstall(1) failure running diff(1) no longer reports "files are different"

== 2.9: December 22, 2023

 - Add flag '-E' for overriding remote environment variables
 - rset(1)/renv(1) extended to store session-level state using name="value"
   arguments
 - renv(1) improved error handling, one variable per line
 - Immediately raise errors and abort if environment or environment_file options
   contain syntax errors
 - rinstall(1) downloads sources to the staging directory without a temporary
   file to avoid duplication
 - rinstall(1) prioritize curl with fallback to wget on Linux
 - rinstall(1) recreates file path when fetching over HTTP
 - rinstall(1) allows empty files to be installed
 - Only support '-l' flag with dry-run
 - New custom log format for each stage of execution defined by environment
   variables

== 2.8: November 13, 2023

 - No longer set ROUTE_LABEL, and LABEL on remote host
 - New renv(1) utility for validating environment variables
 - New environment_file option for defining variables
 - New environment option for defining variables inline
 - Improve parsing by prioritizing parameters before labels
 - '-l' requires an option value

== 2.7: October 28, 2023

 - Avoid opening temporary files twice
 - Connect to the hostname specified, not the primary label name
 - Do not start web server during dry run
 - Execute all hostnames specified on the command line
 - Support for hostname sequence ranges in routes file

== 2.6: September 6, 2023

 - rinstall: use wget with fallback to curl except for *BSD
 - configure: use TARGET_OS to override the output of uname(1)
 - MacOS: use tar -c --no-xattrs
 - Use mkstemp(3) to avoid compiler warnings
 - Add bundler/inline and install webrick locally to run tests

== 2.5: December 16, 2022

 - rinstall: ignore environment variable http_proxy
 - rinstall: Use shell getopts to parse arguments
 - Raise error if a label alias begins with a space

== 2.4: May 3, 2022

 - Formatting improvements to tests
 - Route labels may be a comma-separated list

== 2.3: November 17, 2021

 - Tweaks to unveil(2)
 - rinstall(1) accepts a directory instead of a target, similar to cp(1)

== 2.2: June 16, 2021

 - Raise an error if the output of local execution does not end with '\n'
 - Rename "Remote Sequential Execution Tool" to "Remote Staging Execution Tool"
 - Fix miniquark arg parsing on arm

== 2.1: April 3, 2021


rsub.sh
#!/bin/sh
# rsub
# A helper utility for rset(1)
# Substitute lines in a file or append if not found

set_defaults() {
	ret=1     # global exit status
	append=1  # set to 0 to append text in line-replace mode
	line_regex=""
	line_text=""
	: ${RSUB_START:="# start managed block"}
	: ${RSUB_END:="# end managed block"}
	: ${RSUB_DIFF_ARGS:="-U 2"}
}

main() {
	set_defaults
	parse_args "$@"
	init

	if [ -z "$line_regex$line_text" ]; then
		replace_block
	else
		replace_line
	fi

	# Source file exists: check the difference between source and target
	check_diff_source_target

	rm -f $source
	exit $ret
}

usage() {
	>&2 echo "release: ${release}"
	>&2 echo "usage: rsub [-A] -l line_text -r line_regex target"
	>&2 echo "usage: rsub target < block_content"
	if [ -z "$1" ]; then
		echo >&2 "hint: use -h to display option summary"
		exit 1
	fi

	cat <<- HELP
		summary:
		    -A             Append line if the pattern is not found
		    -l line_text   Replacement text for a matching line
		    -r line_regex  Regular expression matching the line to replace
		docs:
		    man rsub
	HELP
	exit 1
}

init() {
	if ! check_absolute_path "$target"; then
		>&2 echo "rsub: $target is not an absolute path"
		exit 1
	fi

	[ -f "$target" ] || {
		>&2 echo "rsub: file not found: $target"
		exit 3
	}

	source=$(mktemp rsub_XXXXXXXX)

	trap '' HUP
}

parse_args() {
	[ "x$1" = "x-h" ] && usage $1
	while getopts Al:r: arg; do
		case "$arg" in
			A) append=0 ;;
			r) line_regex="$OPTARG" ;;
			l) line_text="$OPTARG" ;;
			?) usage ;;
		esac
	done
	shift $(($OPTIND - 1))
	[ $# -eq 1 ] || usage
	target=$1
}

check_diff_source_target() {
	[ -e "$target" ] && diff $RSUB_DIFF_ARGS "$target" $source || {
		cp $source "$target"
		ret=0
	}
}

replace_block() {
	cat <<-EOF > ${source}_b
	${RSUB_START}
	`cat`
	${RSUB_END}
	EOF
	awk -v m="$RSUB_START" -v n="$RSUB_END" -v block=${source}_b '
	    $0 == m, $0 == n { if (!x) system("cat " block) ; x=1; next; }; 1
	    END { if (!x) system("cat " block) }
	    ' "$target" > $source
	rm ${source}_b
}

replace_line() {
	awk -v n=$append -v a="$line_regex" -v b="$line_text" '
	    BEGIN { gsub("&", "\\\\&", b) }
	    { n+=sub(a, b); print }
	    END { if (n==0) print b }
	    ' "$target" > $source
}

check_absolute_path() {
	case $(dirname "$1") in
		/*) return 0
			;;
		*)  return 1
			;;
	esac
}

main "$@"
